crates/apollo-router-core/src/query_planner/model.rs

```diff
@@ -70,6 +75,185 @@ impl PlanNode {
             Self::Flatten(flatten) => Box::new(flatten.node.service_usage()),
         }
     }
+
+    /// Recursively validate a query plan node making sure that all services are known before we go
+    /// for execution.
+    ///
+    /// This simplifies processing later as we can always guarantee that services are configured for
+    /// the plan.
+    ///
+    /// # Arguments
+    ///
+    ///  *   `plan`: The root query plan node to validate.
+    pub fn validate_services_against_plan(
+        &self,
+        service_registry: Arc<dyn ServiceRegistry>,
+    ) -> Vec<FetchError> {
+        self.service_usage()
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .filter(|service| !service_registry.has(service))
+            .map(|service| FetchError::ValidationUnknownServiceError {
+                service: service.to_string(),
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Recursively validate a query plan node making sure that all variable usages are known before we
+    /// go for execution.
+    ///
+    /// This simplifies processing later as we can always guarantee that the variable usages are
+    /// available for the plan.
+    ///
+    /// # Arguments
+    ///
+    ///  *   `plan`: The root query plan node to validate.
+    pub fn validate_request_variables_against_plan(&self, request: &Request) -> Vec<FetchError> {
+        let required = self.variable_usage().collect::<HashSet<_>>();
+        let provided = request
+            .variables
+            .as_ref()
+            .map(|v| v.keys().map(|x| x.as_str()).collect::<HashSet<_>>())
+            .unwrap_or_default();
+        required
+            .difference(&provided)
+            .map(|x| FetchError::ValidationMissingVariable {
+                name: x.to_string(),
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Validate the entire request for variables and services used.
+    #[tracing::instrument(name = "validation")]
+    pub fn validate_request(
+        &self,
+        request: &Request,
+        service_registry: Arc<dyn ServiceRegistry>,
+    ) -> Result<(), Response> {
+        let mut early_errors = Vec::new();
+        for err in self.validate_services_against_plan(Arc::clone(&service_registry)) {
+            early_errors.push(err.to_graphql_error(None));
+        }
+
+        for err in self.validate_request_variables_against_plan(request) {
+            early_errors.push(err.to_graphql_error(None));
+        }
+
+        if !early_errors.is_empty() {
+            Err(Response::builder().errors(early_errors).build())
+        } else {
+            Ok(())
+        }
+    }
+
+    fn execute_inner<'a>(
```

Comment:
it makes much more sense to have this here :+1:

Reply:
ikr! :laughing: